Drone/drone_base.py

```diff
@@ -0,0 +1,361 @@
+
+    def move(self, direction, distance, stop_event, velocity=c.DEFAULT_VELOCITY):
+        """
+        Moves the drone along a path.
+
+        Parameters
+        ----------
+        direction: UP, DOWN, LEFT, RIGHT, FORWARD, BACK (as defined in constants.py) 
+            The direction the drone should travel in
+        distance: Double
+            The distance in meters the drone should travel in the given direction
+        emergency_land: threading.Event()
+            Set when this movement needs to stop
+
+        Precondition:
+        ----------
+        The drone is already in the air and not moving anywhere else
+
+        Postcondition:
+        ----------
+        The drone has moved in the specified direction by the specified number of meters
+
+        Returns:
+        ----------
+        threading.Event
+        """
+
+        # Calculate duration to send velocity command based on distance and velocity
+        duration = int(distance / velocity)
+
+        # Multiply unit vector in direction by the velocity
+        vector = tuple(velocity * n for n in direction)
+
+        # Make the mavlink message
+        msg = dkw.get_velocity_message(self.vehicle.message_factory, vector)
+
+        def move_thread(msg, duration, finished, stop_event):
+            self.logger.info(threading.current_thread().name + ": Starting move")
+            # Send the message once every second
+            for _ in range(0, duration):
+                self.vehicle.send_mavlink(msg)
+                time.sleep(1)
+                if stop_event.is_set():
+                    self.logger.info(threading.current_thread().name + ": Movement halting")
+                    break
+
+            self.logger.info(threading.current_thread().name + ": Finished move")
+            finished.set()
+
+        finished = threading.Event()
+        threading.Thread(target=move_thread, name="MovementThread-" + str(DroneBase.cnt), 
+            args=(msg, duration, finished, stop_event)).start()
+        DroneBase.cnt += 1
+        return finished
+            
+    
+    def hover(self, duration, stop_event):
+        """
+        Causes the drone to hover in the air.
+
+        Parameters
+        ----------
+        duration: Integer
+            The number of seconds that the drone should hover in place
+        stop_event: threading.Event()
+            Set whenever the hover should stop
+
+        Precondition:
+        ----------
+        The drone is flying.
+
+        Postcondition:
+        ----------
+        None
+
+        Returns:
+        ----------
+        None
+        """
+        # Make the mavlink message
+        msg = dkw.get_velocity_message(self.vehicle.message_factory, (0, 0, 0))
+
+        def hover_thread(msg, duration, finished, stop_event):
+            self.logger.info(threading.current_thread().name + ": Starting hover")
+            # Send the message once every second
+            for _ in range(0, duration):
```

Comment:
This code is repeated again from the move_thread function and violates DRY, consider using a Timer class a shown above to schedule events.

Reply:
I believe I have fixed the DRY problems with hover() and with move(). How does this look:
```python
    def hover(self, duration, stop_event,
        send_delay=c.MAV_VELOCITY_MSG_RESEND_DELAY):
        # Make the mavlink message
        msg = dkw.get_velocity_message(self.vehicle.message_factory, (0, 0, 0))

        def hover_thread(msg, duration, finished, stop_event,
            delay=send_delay):
            self.logger.info('{}: Starting hover'.format(current_thread_name()))
            # Send the message once every delay seconds
            timer = Timer()
            timer.add_callback(
                "move_msg", delay, 
                lambda : self.vehicle.send_mavlink(msg),
                recurring=True)

            if stop_event.wait(duration):
                msg = 'Hover halting'
            else:
                msg = 'Hover finished'
            
            timer.stop_callback("move_msg")
            self.logger.info(
                '{}: {}'.format(current_thread_name(), msg))
            
            finished.set()

        finished = threading.Event()
        thread_name = 'HoverThread-{}'.format(str(DroneBase.cnt))
        threading.Thread(
            target=hover_thread, name=thread_name, 
            args=(msg, duration, finished, stop_event)
            ).start()
        DroneBase.cnt += 1
        return finished
```

Reply:
Looks alright to me.